### DISCUSSION TITLE
How to Initialize an Empty Subdomain Block

### DISCUSSION THREAD

**Discussion by czadeh** [idaholab/moose#19084] [Q&A Meshing] [answered]
Hi,

I just posted the other day but I have a different question.

In a heat conduction app I am developing, I am trying to track the interface of a melt pool.

To achieve this I tried to divide the domain into solid and liquid domains. 

I used SubdomainBoundingBoxGenerator to bound the entire domain and set it to block 1 to represent the solid phase.

I plan to use block 2 to represent the liquid domain, but I do not initialize it in the [Mesh] section of the input file because there is no liquid domain initially. 

I set up a system with CoupledVarThresholdElementSubdomainModifier to set the subdomain to block 2 if conditions are met to make it liquid. 

But I run into the error 

>the following blocks (ids) do not exist on the mesh: 2

or 

>Requested subdomain 2 does not exist

My input file only works if subdomain 2 has been declared already, however in my problem liquid does not exist initially.

How do I initialize subdomain 2 as an empty set and then add to it later?

Alternatively, is this the right way to approach this problem with a moving interface or do you recommend another method?

Thanks,
Cameron

**Comment by GiudGiud** [comment, marked as answer]
@dewenyushu @hugary1995 @ngrilli @jiangwen84 @dschwen 
did you come across this / think about this when working on element deletion? I know it s sort of the opposite problem

**Reply by dewenyushu** [reply]
Yes I had similar issues before. I tried to have empty subdomain work but did not end up getting it work well. My work around was to create a small enough subdomain 2 initially (containing only a few elements) instead of leaving it totally empty.

I remember @hugary1995 also met the similar issue and he might have a better solution.

**Reply by czadeh** [reply]
> My work around was to create a small enough subdomain 2 initially (containing only a few elements) instead of leaving it totally empty.

@dewenyushu How did you create that small subdomain? Did you use SubdomainBoundingBoxGenerator to place a smaller box within the bigger box? Or do you somehow have to "cut out" that smaller box from the bigger box?

**Reply by GiudGiud** [reply]
Could you just place this box completely outside of the mesh you care about, then delete it in the first timestep with adaptivity?

**Reply by czadeh** [reply]
I can try that, but what do you mean by delete with adaptivity?

**Reply by GiudGiud** [reply]
See adaptivity
https://mooseframework.inl.gov/syntax/Adaptivity/index.html
deleting elements is done in some mesh generators. Should be fairly easy to add it to adaptivity if it s not already there

**Reply by dewenyushu** [reply]
> > My work around was to create a small enough subdomain 2 initially (containing only a few elements) instead of leaving it totally empty.
> 
> @dewenyushu How did you create that small subdomain? Did you use SubdomainBoundingBoxGenerator to place a smaller box within the bigger box? Or do you somehow have to "cut out" that smaller box from the bigger box?

Yes, I used the SubdomainBoundingBoxGenerator to create a smaller box. I did not cut it out from the bigger box.

**Comment by hugary1995** [comment]
I asked about it but @loganharbour shut me down. Logan, do you think supporting dummy subdomain names that initially contain zero elements in the map is possible in MOOSE? This is also helpful for #19074 .

On the other side, I believe exodus already supports empty blocks. So I think we can get it to work as long as MOOSE can support it. For example, we can provide a mesh generator that creates empty subdomains.

**Reply by loganharbour** [reply]
I don't think this is an awful idea - but we will have to ping @roystgnr.

The issue is that the concept of available subdomains comes from looping through elements and collecting the set of the subdomains set on elements. There are quite a few ways go to about this... but it will require changing things at a very fundamental level in libmesh.

**Reply by roystgnr** [reply]
Well, in libMesh we already do have two distinct concepts here: `MeshBase::get_mesh_subdomains()` tries to return the set of subdomain ids that currently exist in the mesh, and there's no way to change that without breaking the API for people, but `MeshBase::get_subdomain_name_map()` returns all subdomain ids that have been given names, and there's no requirement for any correspondence between the two; you can add an additional id->name entry whether that id yet exists in the mesh or not.

I'm totally unsure about whether we correctly read or write empty blocks to Exodus, though.  We certainly don't have any unit tests checking it though, and it's not the sort of thing we'd have been likely to get right on accident.

**Reply by jmeier** [reply]
Dear @hugary1995, @loganharbour and @roystgnr,

Please excuse me for resurrecting this old (resolved) discussion. In the meantime, is there a way to allow empty subdomains?

I'm trying to build a model that consists of a number of blocks that should be switched off and on individually. The CoupledVarThresholdElementSubdomainModifier is to be used for this. Without "empty blocks" I have several problems:
- The (initially empty) block for the inactive domain must have a few dummy elements as described above.
- If I use CoupledVarThresholdElementSubdomainModifier to make a whole block inactive, it is currently deleted by Moose (and Moose complains that the blocks are no longer numbered consecutively).

The availability of "empty blocks" would be a great help.

Kind regards,
Jörg

**Reply by GiudGiud** [reply]
Hello

yes there is. It s in the Mesh block, you can declare future subdomains that are not in use yet
The `add_subdomain_ids` parameter

Guillaume

**Reply by jmeier** [reply]
Dear @GiudGiud, 

Thank you very much for your answer. This option looks great. Is there an similar option to prevent Moose to delete subdomains after becomming empty (e.g. after moving all elements to another block via CoupledVarThresholdElementSubdomainModifier)?

Jörg

**Reply by GiudGiud** [reply]
It s a good question. Did you try using this same Boolean for that purpose?

**Reply by jmeier** [reply]
That doesn't seem to work. Please see below.

In the code below I load the subdomains 1 to 7 from a MSH file. Furthermore, I ask Moose for a subdomain 8 - and I have also added the subdomains 1 to 7 again in add_subdomain_ids, because I do not want Moose to delete them after I have moved all elements with CoupledVarThresholdElementSubdomainModifier to subdomain 8. Moose accepts this code, creates subdomain 8, but deletes subdomains 1 to 7 later anyway.

```
[Mesh]

  [File]
    type = FileMeshGenerator
    file = "stages.msh"   # This gives subdomains 1 to 7
    show_info = true
  []

  add_subdomain_ids = '1 2 3 4 5 6 7 8'   # I dont want 1 to 7 getting deleted later on - and I need subdomain 8

[]
```

The specific error message I get is:
```
Time Step 0, time = 0

Time Step 1, time = 0.2, dt = 0.2

*** ERROR ***
/home/user/projects/moose_bare_tmech/models/stages.i:95: (UserObjects/BlockActivation_StatesCSV/nblock):
    Element block id 8 greater than than total number of blocks in mesh: 7. Blocks should be numbered consecutively.
```

**Reply by GiudGiud** [reply]
This is not from the mesh losing a subdomain.
This is from the PropertyReadFIle user object which does not update the number of subdomains in the mesh.

The solution here is to implement a `meshChanged()` routine to this UO and make it update the number of subdomains. 
OR
query the number of subdomains dynamically instead of setting it at the construction of the UO
